hazelcast/src/test/java/com/hazelcast/test/TestHazelcastInstanceFactory.java

```diff
@@ -296,33 +310,45 @@ public HazelcastInstance getInstance(Address address) {
      * @param instance the instance to terminate
      */
     public void terminate(HazelcastInstance instance) {
-        Address address = getNode(instance).address;
-        terminateInstance(instance);
-        if (isMockNetwork) {
-            registry.removeInstance(address);
+        try {
+            Address address = getNode(instance).address;
+            terminateInstance(instance);
+            if (isMockNetwork) {
+                registry.removeInstance(address);
+            }
+        } finally {
+            TestJmxLeakHelper.checkJmxBeans(instance.getName());
```

Comment:
We have a large number of tests that don't use `TestHazelcastInstanceFactory` at all and call `Hazelcast.newHazelcastInstance()` directly (in combination with `Hazelcast.shutdownAll()` / `HazelcastInstanceFactory.shutdownAll()` / `HazelcastInstanceFactory.terminateAll()`). Shouldn't those be checked as well, somehow?

Reply:
I think it is possible to do this somewhere here: https://github.com/hazelcast/hazelcast/blob/86a759ac20eab87d7a249fa6c267a5f5ab4ad215/hazelcast/src/test/java/com/hazelcast/test/AbstractHazelcastClassRunner.java#L304-L310

I would prefer that for the reasons Vojtech mentioned. For example, the client tests rarely use `TestHazelcastFactory`, but clients also create MBeans under the same `com.hazelcast.*` namespace which this solution ignores.

Reply:
@vojtechtoman @blazember Thank you very much for suggestions. I implemented the proposed strategy and it helped to identify the leaking class immediately - it was `MetricsServiceTest`. Changing it to serial mode along with explicit JMX cleanup seems to solve the original problem. Let's wait for Jenkins results.